### PR TITLE
Implement websocket (ws:// and wss://) links

### PR DESCRIPTION
**Pull request by basilgello**
ws:// can be listened and dialed
wss:// is a convenience link for ws:// that supports dialing to ws:// peer.

### CODE REVIEW

File: src/core/link_ws.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+type linkWSS struct {
+	*linkWS
+	tlsconfig  *tls.Config
+}
+
+type linkWSConn struct {
+	net.Conn
+}
+
+type linkWSListener struct {
```
Comment (neilalexander) [inline src/core/link_ws.go:29]: Looks like this file isn't `gofmt`'d?

File: src/core/link_ws.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+func (l *linkWSListener) Addr() net.Addr {
+	return l.listener.Addr()
+}
+
+func (l *linkWSSListener) Addr() net.Addr {
+	return l.listener.Addr()
+}
+
+func (l *linkWSListener) Close() error {
```
Comment (neilalexander) [inline src/core/link_ws.go:73]: Can shorten to something like:
```
{
  if err := l.httpServer.Shutdown(l.ctx); err != nil {
    return err
  }
  return l.listener.Close()
}
```

File: src/core/link_ws.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+	if err != nil {
+		return err
+	}
+
+	err = l.listener.Close()
+
+	return err
+}
+
+func (l *linkWSSListener) Close() error {
```
Comment (neilalexander) [inline src/core/link_ws.go:87]: Likewise here:
```
{
  if err := l.httpServer.Shutdown(l.ctx); err != nil {
    return err
  }
  if err := l.tlslistener.Close(); err != nil {
    return err
  }
  return l.listener.Close()
}
```

File: src/core/link_ws.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+	return err
+}
+
+func (s *wsServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	c, err := websocket.Accept(w, r, &websocket.AcceptOptions{
+		Subprotocols: []string{"ygg-ws"},
+	})
+
+	if err != nil {
+		return
```
Comment (neilalexander) [inline src/core/link_ws.go:113]: Do we want to return an error code in this situation? i.e.
```
  if err != nil {
    w.WriteHeader(http.StatusUpgradeRequired)
    return
  }
```
... or is the WebSocket library doing that automatically for us?

**Review by neilalexander — CHANGES_REQUESTED** [review]
Principally fine, might be best to split into `link_ws.go` and `link_wss.go` for brevity/clarity, but a few inline comments below.

File: src/core/link_ws.go (hunk 1 above)
Comment (basilgello) [reply src/core/link_ws.go:29]: Yup, my fault :)

File: src/core/link_ws.go (hunk 2 above)
Comment (basilgello) [reply src/core/link_ws.go:73]: Done

File: src/core/link_ws.go (hunk 3 above)
Comment (basilgello) [reply src/core/link_ws.go:87]: Done

File: src/core/link_ws.go (hunk 4 above)
Comment (basilgello) [reply src/core/link_ws.go:113]: Yes, it does one for us and this is mentioned in author's "production quality demo project" I used as source

File: src/core/link_ws.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	lwss := &linkWSS{
+		linkWS:    lws,
+		tlsconfig: l.core.config.tls.Clone(),
+	}
+
+	return lwss
+}
+
+func (l *linkWS) dial(ctx context.Context, url *url.URL, info linkInfo, options linkOptions) (net.Conn, error) {
+	wsconn, _, err := websocket.Dial(ctx, url.Scheme + "://" + url.Host, &websocket.DialOptions{
```
[the remark this answers is not in the document]
Comment (basilgello) [reply src/core/link_ws.go:149]: Makes sense

**Comment by basilgello** [comment]
LGTM :)

**Pull request merged**